### PR TITLE
SpinButton: widen `value` type to include `null`

### PR DESCRIPTION
**Pull request by spmonahan**
## Current Behavior

`value` and `defaultValue` are typed as the TS `number` type.

## New Behavior

`value` and `defaultValue` have their types widened to `number | null`.

This change is being made based on partner feedback for wanting a reliable way to represent "no value" or "unset value".

`null` was chosen because Typescript can easily ensure users handle the null case as opposed to a value like `NaN`. `null` is also obviously a special case so it's clearer that it should be handled specially. Because `null` is a special, non-`number` value it can more simply be handled internally by SpinButton.

#### undefined

`undefined` cannot be used as React will interpret `value`/`defaultValue` toggling from a number to `undefined` and back as the component switching from controlled to uncontrolled (which is logged as an error).

React error message about toggling between uncontrolled and controlled component

#### NaN

[NaN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NaN) stands for "not a number" yet it is a number both to Javascript and Typescript:

```ts
console.log(typeof NaN); // "number"
const aNumber: number = NaN; // No TS warnings or errors
```

On the one hand this is convenient as the types for `value` and `defaultValue` could remain simply `number`, however, there are some issues when working with `NaN`:

1. We'd be using `NaN` as a special case yet it's a perfectly valid number and TS will not issue a warning if users don't explicitly handle it. Since it would be part of SpinButton's API it's important that users handle this case. With `number | null` TS will warn users if they don't handle the `null` case.
2. `NaN == NaN` and `NaN === NaN` are both `false` so users must use either the global `isNaN()` function or `Number.isNaN()`. In addition to the friction of not being able to use equality operators on a number value, users now have to make another decision around using `isNaN()` vs `Number.isNaN()` as they have different behavior.

#### Leave the type as `number` and use "special" numbers

Users could choose an "unset" value that is also a number like `-1` but this only works if the "unset" value is outside the [min, max] SpinButton range (and won't work if there is no [min, max] range). Additionally, since SpinButton would have no way of knowing about this special value users would have to implement special handling for scenarios like `value` being set to the "special" number when the user presses the increment button.

## Related Issue(s)

#22673

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                      |     Baseline (minified/GZIP) |                           PR |                                                                                                                                                                                                  Change |
| :----------------------------------------------------------------------------------------------------- | ---------------------------: | ---------------------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-spinbutton</samp> <br /> <abbr title='bundle-size/SpinButton.fixture.js'>SpinButton</abbr> | `41.985 kB`<br />`11.908 kB` | `42.152 kB`<br />`11.954 kB` | `167 B` <br />`46 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |         Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `179.813 kB`<br />`50.19 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            | `34.178 kB`<br />`11.138 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/73f592de5bace4132289e64b1df78e9cc4b31a25'>73f592de5bace4132289e64b1df78e9cc4b31a25</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/Avatar-mount_base.html">1117</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/Avatar-mount.html">1060</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/Button-mount_base.html">682</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/Button-mount.html">646</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/FluentProvider-mount_base.html">2182</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/FluentProvider-mount.html">2217</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/FluentProviderWithTheme-mount_base.html">324</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/FluentProviderWithTheme-mount.html">338</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">285</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">285</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">373</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">357</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/MakeStyles-mount_base.html">1901</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23025/perf-test-react-components/MakeStyles-mount.html">1876</a></td>
            <td>50000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.types.ts [hunk 1]
```diff
@@ -32,7 +32,10 @@ export type SpinButtonSlots = {
 /**
  * SpinButton Props
  */
-export type SpinButtonProps = Omit<ComponentProps<Partial<SpinButtonSlots>, 'input'>, 'onChange' | 'size'> & {
+export type SpinButtonProps = Omit<
+  ComponentProps<Partial<SpinButtonSlots>, 'input'>,
+  'onChange' | 'size' | 'value' | 'defaultValue'
```
Comment (khmakoto) [inline packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.types.ts:37]: _nit:_ Keep these ordered

```suggestion
  'defaultValue' | 'onChange' | 'size' | 'value'
```

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx [hunk 2]
```diff
@@ -47,7 +47,7 @@ export const useSpinButton_unstable = (props: SpinButtonProps, ref: React.Ref<HT
   const nativeProps = getPartitionedNativeProps({
     props,
     primarySlotTagName: 'input',
-    excludedPropNames: ['onChange', 'size', 'min', 'max'],
+    excludedPropNames: ['onChange', 'size', 'min', 'max', 'value', 'defaultValue'],
```
Comment (khmakoto) [inline packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:50]: _nit:_ Order these alphabetically

```suggestion
    excludedPropNames: ['defaultValue', 'max', 'min', 'onChange', 'size', 'value'],
```

**Review by khmakoto — APPROVED** [review]

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx [hunk 3]
```diff
@@ -266,13 +275,15 @@ export const useSpinButton_unstable = (props: SpinButtonProps, ref: React.Ref<HT
 
     if (valueChanged || displayValueChanged) {
       onChange?.(e, { value: roundedValue, displayValue: newDisplayValue });
+    } else if (!valueChanged && newValue === null) {
+      onChange?.(e, { value: null });
     }
```
Comment (jspurlin) [inline packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:280]: Will this fire on all blurs, ENTER keydown, any attempts to step (if so, is that what we want)?

File: packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx [hunk 4]
```diff
@@ -260,6 +278,39 @@ describe('SpinButton', () => {
     expect(onChange.mock.calls[1][1]).toEqual({ value: 2, displayValue: undefined });
   });
 
+  it('calls on change when defaultValue is `null` when uncontrolled', () => {
```
Comment (jspurlin) [inline packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx:281]: > it('calls on change when defaultValue is `null` when uncontrolled' [](http://example.com/codeflow?start=1&length=67)

Consider potentially adding a tests the number pf onChange calls for multiple steps/blurs/ENTERs when the value is null (e.g. `expect(onChange).toHaveBeenCalledTimes(x);`) to explicitly keep track of how onChange is expected to get called in this case. I feel like if there's not a test like this, then the implicitly expected number of times onChange gets called in this case could change unexpectedly

(this is similar to my other comment in `useSpinButton`)

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx (hunk 3 above)
Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:280]: It does and it was intentional. My thinking was that `null` is a special case so SpinButton should fire events when a user tries to change it but thinking about it more, I'm not so sure. `null` has to be set from the app, SpinButton never sets the value to `null` internally, so the app will be fully aware of this state. I'm thinking I'll remove this special case.

What do you think?

Comment (jspurlin) [reply packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:280]: I don't have a strong opinion, but my gut instinct is that it'd be nice if the firing of `onChange` was consistent

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx [hunk 5]
```diff
@@ -266,13 +275,15 @@ export const useSpinButton_unstable = (props: SpinButtonProps, ref: React.Ref<HT
 
     if (valueChanged || displayValueChanged) {
       onChange?.(e, { value: roundedValue, displayValue: newDisplayValue });
+    } else if (!valueChanged && newValue === null) {
+      onChange?.(e, { value: null });
```
Comment (jspurlin) [inline packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:279]: >{ value: null } [](http://example.com/codeflow?start=20&length=15)

Any thoughts on making the shape of the data returned consistent (e.g. make the object always have both `value` and `newDisplayValue`)? I think doing that could potentially allow consumers to simplify logic in their `onChange` callback

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx (hunk 3 above)
Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:280]: Agreed. Based on our conversation [here](https://github.com/microsoft/fluentui/pull/23004#discussion_r874289988) I'm going to update this so it only fires when the value actually changes.

File: packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx [hunk 6]
```diff
@@ -320,6 +348,97 @@ describe('SpinButton', () => {
...
+
+    expect(onChange).not.toHaveBeenCalled();
+    expect(getSpinButtonInput().value).toBe('');
+
+    rerender(<SpinButton value={null} onChange={onChange} />);
+    userEvent.click(decrementButton);
+
+    expect(onChange).not.toHaveBeenCalled();
+    expect(getSpinButtonInput().value).toBe('');
   });
```
Comment (jspurlin) [inline packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx:442]: How does a user get out of a null value?

I'm considering the case where the SpinButton is empty and the user doesn't know what to type into the SpinButton manually so they try stepping instead... It looks like nothing will happen now?

When in this case, how would the creator know that the user is trying to set the SpinButton unless the explicitly type into the SpinButton?

I feel like it should fire on increment/decrement, maybe with the value of 1/-1 respectively? That would allow the user to to get out of the null case by stepping in the uncontrolled case, while also allowing the creator to know what the user's intentions are (at that point it's up to the creator to change it to something non-null, or potentially be notified again with a 1/-1 value).

For non-step cases (e.g. ENTER/blur) I would expect onChange to not fire if the value remained empty

Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx:442]: > I'm considering the case where the SpinButton is empty and the user doesn't know what to type into the SpinButton manually (or the SpinButton is non-editable, (e.g. does not allow freeform input)... is that even possible today, maybe the creator could pass readonly to the input???) so they try stepping instead... It looks like nothing will happen now?

An author could use slots to disable the input but we've decided against disallowing freeform input as part of SpinButton's API (it was there previously) because it makes the control inaccessible (there's no indication the input field disallows input). Slots give authors a lot of power when using components and I'm hesitant to make changes to the built-in behavior just because someone could do something. In this scenario, I think my response is, "don't disable the input as you're making the SpinButton inaccessible".

> When in this case, how would the creator know that the user is trying to set the SpinButton unless the explicitly type into the SpinButton?

Today you'd have to use the slots API to attach listeners to the input field, increment button and decrement button. I've opened a [feature request about an "intent to change" event handler](https://github.com/microsoft/fluentui/issues/23043)

> I feel like it should fire on increment/decrement, maybe with the value of 1 / -1 respectively? That would allow the user to to get out of the null case by stepping in the uncontrolled case, while also allowing the creator to know what the user's intentions are (at that point it's up to the creator to change it to something non-null, or potentially be notified again with a 1 / -1 value).Note, that is if no min is provided, if a min is provided you could return min+1 / min

Would this apply to the uncontrolled case only?

Since the only way for SpinButton to have its value set to `null` is for it to be set via props (SpinButton will never set null internally) I wonder if it's better to remove `| null` from the type of `defaultValue`?

File: packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx [hunk 7]
```diff
@@ -318,6 +346,97 @@ describe('SpinButton', () => {
...
+
+  it('calls on change when value is `null` when controlled', () => {
+    const onChange = jest.fn();
+    const { getAllByRole, rerender } = render(<SpinButton value={null} min={5} max={10} onChange={onChange} />);
+
+    const [incrementButton, decrementButton] = getAllByRole('button');
+    userEvent.click(incrementButton);
+
+    expect(onChange).toHaveBeenCalledTimes(1);
+    expect(onChange.mock.calls[0][1]).toEqual({ value: 5, displayValue: undefined });
```
Comment (smhigley) [inline packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx:433]: Since the assumed starting value here is 5, shouldn't `onChange` be called with `value: 6` after the incrementButton click?

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx [hunk 8]
```diff
@@ -163,9 +166,15 @@ export const useSpinButton_unstable = (props: SpinButtonProps, ref: React.Ref<HT
   };
 
   const stepValue = (e: SpinButtonChangeEvent, direction: 'up' | 'down' | 'upPage' | 'downPage') => {
+    const val = internalState.current.value;
     const dir = direction === 'up' || direction === 'upPage' ? 1 : -1;
     const stepSize = direction === 'upPage' || direction === 'downPage' ? stepPage : step;
-    const val = internalState.current.value;
+
+    if (val === null) {
+      const nullStep = clamp(0 + stepSize * dir, min, max);
```
Comment (smhigley) [inline packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:174]: Should 0 be hardcoded here, or should we default to starting from the min?

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx [hunk 9]
```diff
@@ -272,7 +282,7 @@ export const useSpinButton_unstable = (props: SpinButtonProps, ref: React.Ref<HT
   state.input.value = textValue;
   state.input['aria-valuemin'] = min;
   state.input['aria-valuemax'] = max;
-  state.input['aria-valuenow'] = currentValue;
+  state.input['aria-valuenow'] = currentValue === null ? undefined : currentValue;
```
Comment (smhigley) [inline packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:285]: minor nit: although it seems weird, could this be shortened to `currentValue ?? undefined`?

File: packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx (hunk 7 above)
Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx:433]: I think you can make an argument for that. This behavior is based on the `input type="number"` which you can observe here: https://codepen.io/seanms/pen/BaYZVrL?editors=1010

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx (hunk 8 above)
Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:174]: The `0` can obviously be removed but I felt leaving makes it clearer where the step is starting from and leaving it has no material impact on performance.

`min` and `max` aren't necessarily both defined. You can define neither, both or only one and `clamp()` takes this into account. The only constraints are that they are both numbers and `min <= max`. `max` can be less than 0 so it's possible the value would be clamped to `max` rather than `min`.

I'll update tests to reflect this.

File: packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx (hunk 9 above)
Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/components/SpinButton/useSpinButton.tsx:285]: Maybe? I would have sworn that `null` would render `aria-valuenow=""` but that does not seem to be the case :)

File: packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx (hunk 6 above)
Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx:472]: I spent some time thinking this over and I agree with your assessment. I've updated the null behavior to work with increment/decrement buttons.

**Review by smhigley — APPROVED** [review]
